### PR TITLE
Feature/add product

### PR DESCRIPTION
**Pull request by khai-pi**
Add product feature for backend

### CODE REVIEW

File: src/business/products/add-products.ts [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    type: "/products/add-product-failed",
+    message: "product name exist",
+  },
+};
+
+const AddProductRequest = Type.Object({
+  name: Type.String(),
+  sku: Type.String(),
+  price: Type.String(),
+  stock_quantity: Type.Union([ Type.Number() ], Type.Null()), // string || null
```
Comment (tungUAS) [inline src/business/products/add-products.ts:55]:
```suggestion
  stock_quantity: Type.Optional( Type.Number() )
```

File: src/business/products/add-products.ts [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    userFoundInFirestore.woo_credentials.token,
+    userFoundInFirestore.woo_credentials.secret,
+  );
+
+  const base_url =
+    process.env["NOVE_ENV"] === "product"
+      ? userFoundInFirestore.store.app_url
+      : (process.env["WOO_BASE_URL"] as string);
+
+  try {
```
Comment (tungUAS) [inline src/business/products/add-products.ts:115]: why try catch? we wrap the function in try - catch already. look 
[https://github.com/Woo-Pick-Fullfillment-Inventory/Backend-Woo-Inventory/pull/53/files#diff-64fd237c06fdb42ecf1266b0e420cd4216918363f27b8af8afa425361adb2578R12)]

File: src/modules/create-axios-client.ts [hunk 3]
```diff
@@ -23,10 +24,18 @@ const createAxiosClient = <T>({
...
+      interceptor.onTrue,
+      interceptor.onError,
+    );
   });
 
-  return { get: (url: string, config?: AxiosRequestConfig) => axiosClient.get<T, AxiosResponse<T>>(url, config) };
+  return {
+    get: (url: string, config?: AxiosRequestConfig) =>
+      axiosClient.get<T, AxiosResponse<T>>(url, config),
+    post: (url: string, data: NewProductType, config?: AxiosRequestConfig) =>
```
[the remark this answers is not in the document]
Comment (tungUAS) [reply src/modules/create-axios-client.ts:36]: have a look at get-methode. the post should be generics, compatible for all input types.

File: src/modules/create-axios-client.ts [hunk 4]
```diff
@@ -23,10 +24,18 @@ const createAxiosClient = <T>({
...
+      axiosClient.post<T, AxiosResponse<T>>(url, data, config),
+  };
```
Comment (tungUAS) [inline src/modules/create-axios-client.ts:38]:
```suggestion
      post: <D>( url: string, data: D, config?: AxiosRequestConfig ) => instance.post<R, AxiosResponse<R>>(
        url,
        data,
        config
      ),
```

Comment (tungUAS) [reply src/modules/create-axios-client.ts:38]: this is a module. NewProductType is case specific. go for generics type

File: src/repository/woo-api/create-post-product.ts [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+              `onError Intercepted: request ${error.config.url}:`,
+              error,
+            );
+          }
+          return error;
+        },
+      },
+    ],
+  });
+  const { data } = await post(
```
Comment (tungUAS) [inline src/repository/woo-api/create-post-product.ts:66]: are we really expecting data from Woo API or just a 200 status code?

File: src/repository/woo-api/create-post-product.ts [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+      baseURL: baseUrl,
+      headers: {
+        "Content-Type": "application/json",
+        Authorization: token,
+      },
+    },
+    interceptors: [
+      {
+        onTrue: (response: AxiosResponse) => {
+          if (response.status !== 200) {
```
Comment (tungUAS) [inline src/repository/woo-api/create-post-product.ts:37]: did you test the woo api? post requests often return 201

File: src/repository/woo-api/create-post-product.ts [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+            logger.log(
+              "error",
+              `onTrue Intercepted: request ${response.config.url} with status code ${response.status} is not expected`,
+            );
+            throw new Error("Response not expected");
+          }
+          if (!isResponseTypeTrue(ProductsSchema, response.data, true)) {
```
Comment (tungUAS) [inline src/repository/woo-api/create-post-product.ts:44]: in the post request. there is no need to check for type returned. only status code plays a role.

File: src/repository/woo-api/models/products.type.ts [hunk 8]
```diff
@@ -28,16 +30,39 @@ export const ProductsSchema = Type.Array(ProductSchema);
...
-export type ProductsType = ProductType[];
\ No newline at end of file
+    src: string;
+  }[];
+};
+
+export type ProductsType = ProductType[];
+
+// When user create new product there is no id yet
+export type NewProductType = {
```
Comment (tungUAS) [inline src/repository/woo-api/models/products.type.ts:49]: why the fuck we need this?

File: src/repository/woo-api/create-post-product.ts [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+import type {
+  AxiosError,
+  AxiosResponse,
+} from "axios";
+
+type PostProductResponse = {
+  product: ProductType; // Assuming ProductType is the type for a single product
+};
+
+export const postProducts = async (
```
Comment (tungUAS) [inline src/repository/woo-api/create-post-product.ts:21]:
```suggestion
export const postProduct = async (
```

File: src/business/products/add-products.ts [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+  Request,
+  Response,
+} from "express";
+
+dotenv.config();
+
+const SERVICE_ERRORS = {
+  notAuthorized: {
+    statusCode: StatusCodes.UNAUTHORIZED,
+    type: "/products/not-authorized",
```
Comment (tungUAS) [inline src/business/products/add-products.ts:26]:
```suggestion
    type: "/products/add-product/not-authorized",
```

File: src/repository/woo-api/create-post-product.ts [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+      {
+        onTrue: (response: AxiosResponse) => {
+          if (response.status !== 201) {
+            logger.log(
+              "error",
+              `onTrue Intercepted: request ${response.config.url} with status code ${response.status} is not expected`,
+            );
+            throw new Error("Response not expected");
+          }
+          if (!isResponseTypeTrue(ProductsSchema, response.data, true)) {
```
Comment (tungUAS) [inline src/repository/woo-api/create-post-product.ts:46]:
```suggestion
          if (!isResponseTypeTrue(ProductsSchema, response.data, true).isValid) {
```

File: src/repository/woo-api/models/products.type.ts [hunk 12]
```diff
@@ -33,16 +33,39 @@ export const ProductsSchema = Type.Array(ProductSchema);
...
+    images: {
+      id: number;
+      src: string;
+    }[];
+}
+
+export type ProductsType = ProductType[];
+
+// When user create new product there is no id yet
+export type NewProductType = {
```
Comment (tungUAS) [inline src/repository/woo-api/models/products.type.ts:52]:
```suggestion
export type NewProductType = Omit<ProductType, "id">
```

File: src/repository/woo-api/create-post-product.ts [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+  NewProductType,
+  ProductFromWooType,
+  ProductType,
+} from "./models/products.type.js";
+import type {
+  AxiosError,
+  AxiosResponse,
+} from "axios";
+
+type PostProductResponse = {
```
Comment (tungUAS) [inline src/repository/woo-api/create-post-product.ts:19]: ?

File: src/repository/woo-api/create-post-product.ts [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+      },
+    ],
+  });
+  const { data } = await post(
+    "/wp-json/wc/v3/products",
+    productDetails,
+    { headers: { Authorization: token } },
+  );
+
+  return { product: convertWooProductToClient(data) };
```
Comment (tungUAS) [inline src/repository/woo-api/create-post-product.ts:74]:
```suggestion
  return { convertWooProductToClient(data) };
```

File: src/repository/woo-api/create-post-product.ts [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+
+type PostProductResponse = {
+  product: ProductType; // Assuming ProductType is the type for a single product
+};
+
+export const postProduct = async (
+  baseUrl: string,
+  token: string,
+  productDetails: NewProductType,
+): Promise<PostProductResponse> => {
```
Comment (tungUAS) [inline src/repository/woo-api/create-post-product.ts:27]:
```suggestion
): Promise<ProductType> => {
```

File: src/repository/woo-api/create-post-product.ts [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+} from "axios";
+
+type PostProductResponse = {
+  product: ProductType; // Assuming ProductType is the type for a single product
+};
+
+export const postProduct = async (
+  baseUrl: string,
+  token: string,
+  productDetails: NewProductType,
```
Comment (tungUAS) [inline src/repository/woo-api/create-post-product.ts:26]:
```suggestion
  productDetails: Omit<ProductType,"id">,
```

File: src/business/products/add-products.ts [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+
+  const base_url =
+    process.env["NOVE_ENV"] === "product"
+      ? userFoundInFirestore.store.app_url
+      : (process.env["WOO_BASE_URL"] as string);
+
+  const { product } = await postProduct(
+    base_url,
+    wooBasicAuth,
+    productDetails,
```
Comment (tungUAS) [inline src/business/products/add-products.ts:108]:
```suggestion
req.body```

Comment (tungUAS) [reply src/business/products/add-products.ts:108]: dont declare extra variable if it is used only once. avoid side effects!!

File: src/repository/woo-api/models/products.type.ts (hunk 12 above)
Comment (khaihoang-shopping) [reply src/repository/woo-api/models/products.type.ts]: Keep this to maybe extend response from Woo later

**Comment by khai-pi** [comment]
Keep main branch with some add product function, deal with additional attributes later

File: src/business/products/add-products.ts [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+  const wooBasicAuth = createBasicAuthHeaderToken(
+    userFoundInFirestore.woo_credentials.token,
+    userFoundInFirestore.woo_credentials.secret,
+  );
+
+  const base_url =
+    process.env["NOVE_ENV"] === "product"
+      ? userFoundInFirestore.store.app_url
+      : (process.env["WOO_BASE_URL"] as string);
+
```
Comment (tungUAS) [inline src/business/products/add-products.ts:103]: `process.env["NOVE_ENV"] === "production"`

File: src/repository/woo-api/post-add-product.ts [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+      },
+    ],
+  });
+  const { data } = await post(
+    "/wp-json/wc/v3/products",
+    productDetails,
+    { headers: { Authorization: token } },
+  );
+
+  console.log("=====================================");
```
Comment (tungUAS) [inline src/repository/woo-api/post-add-product.ts:71]: no console log in production

File: src/repository/woo-api/post-add-product.ts [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+} from "./models/products.type.js";
+import type {
+  AxiosError,
+  AxiosResponse,
+} from "axios";
+
+export const postAddProductFactory = async (
+  baseUrl: string,
+  token: string,
+  productDetails: NewProductType,
```
Comment (tungUAS) [inline src/repository/woo-api/post-add-product.ts:22]: direct insert Omit Type here?

**Comment by tungUAS** [comment]
@khai-pi  this PR wasnt ready and cant be merged ? did you test it in production ? pretty sure 500 Code will be submited. please resolve

**Pull request merged**